File: Baekjoon/Class3/45.16928 뱀과 사다리 게임.py
from collections import deque
import sys
input = sys.stdin.readline

BOARD_SIZE = 100

def count_rolls(ladders_and_snakes):
    roll_counts = [0] * (BOARD_SIZE + 1) #암시적 그래프
    bfs = deque([1]) #시작 위치
    while bfs:
        start = bfs.popleft()
        for move in range(1, 7):
            end = start + move
            if end < BOARD_SIZE and ladders_and_snakes[end] != 0: #사다리 뱀 타기
                end = ladders_and_snakes[end]
            if end <= 0 or end > BOARD_SIZE: #경계 검사
                continue
            if roll_counts[end] == 0: #이동
                roll_counts[end] = roll_counts[start] + 1
                if end == BOARD_SIZE:
                    return roll_counts[end]
                bfs.append(end)
    return -1


def main():     
    n, m = map(int, input().split())

    ladders_and_snakes = [0] * BOARD_SIZE
    for _ in range(n + m):
        start, end = map(int, input().split())
        ladders_and_snakes[start] = end

    print(count_rolls(ladders_and_snakes))

main()

File: Baekjoon/Class3/45.16928 뱀과 사다리 게임.js
const inputFile = process.platform === "linux" ? "/dev/stdin" : "./input.txt";
const input = require("fs").readFileSync(inputFile, "utf8").trim().split("\n");

const BOARD_SIZE = 100;

const countRolls = (laddersAndSnakes) => {
    const rollCounts = new Array(BOARD_SIZE + 1).fill(0); // 암시적 그래프
    let bfsHead = 0;
    const bfs = [1]; // 시작 위치
    while(bfsHead < bfs.length) {
        const start = bfs[bfsHead++];
        for(let move = 1; move <= 6; move++) {
            let end = start + move;
            if(end < BOARD_SIZE && laddersAndSnakes[end] !== 0) // 사다리 뱀 타기
                end = laddersAndSnakes[end];
            if(end <= 0 || end > BOARD_SIZE) // 경계 검사
                continue;
            if(rollCounts[end] === 0) { // 이동
                rollCounts[end] = rollCounts[start] + 1;
                if (end === 100)
                    return rollCounts[end];
                bfs.push(end);
            }
        }
    }
}

const [n, m] = input[0].split(" ").map(Number);

const laddersAndSnakes = new Array(BOARD_SIZE).fill(0);
for(let i = 1; i <= n + m; i++) {
    const [start, end] = input[i].split(" ").map(Number);
    laddersAndSnakes[start] = end;
}

console.log(countRolls(laddersAndSnakes));